File: Leetcode/sort_and_search/74_SearchAn2DMatrix/Solution.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix.length==0 || matrix[0].length==0)
            return false;
        int n = matrix[0].length;
        return twoDBisect(matrix, 0, n-1, target);
    }
    private boolean twoDBisect(int[][] matrix, int m, int n, int target){
        if(m==matrix.length || n==-1)
            return false;
        int mid = matrix[m][n];
        return (mid==target)? true: (mid<target)? twoDBisect(matrix, m+1, n, target): twoDBisect(matrix, m, n-1, target);
    }
}

File: Leetcode/sort_and_search/74_SearchAn2DMatrix/Solution.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if not matrix or len(matrix[0])==0:
            return False
        mid = matrix[0][len(matrix[0])-1]
        if mid > target:
            for row in matrix:
                row.pop()
            return self.searchMatrix(matrix, target)
        return True if mid==target else self.searchMatrix(matrix[1:], target)
